### PR TITLE
Gettoken for concurrent calls 

### PR DESCRIPTION
**Pull request by anandrgitnirman**
#309 
Get a token to make concurrent calls

### CODE REVIEW

**Comment by vsbogd** [comment]
I skimmed the PR looks good to me, but better to be reviewed accurately by someone else.

File: escrow/memory_storage.go [hunk 1]
```diff
@@ -110,13 +109,101 @@ func (storage *memoryStorage) Clear() (err error) {
...
+	}
+	return data, false
+}
 func (storage *memoryStorage) CompleteTransaction(transaction Transaction, update []KeyValueData) (ok bool, err error) {
-	return false, fmt.Errorf("Not implemented")
+	originalValues := transaction.(*memoryStorageTransaction).ConditionValues
+	for _, olddata := range originalValues {
+		if olddata.Present {
+			//make sure the current value is the same as the value last read
+			currentValue, _, _ := storage.Get(olddata.Key)
```
Comment (vsbogd) [inline escrow/memory_storage.go:141]: It makes sense to check for `err` and `ok`, `err` should be `nil` and `ok` should be `true`

File: escrow/token_service.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+	channel, ok, err := service.channelService.PaymentChannel(&PaymentChannelKey{ID: channelId})
+
+	if !ok {
+		return fmt.Errorf("channel is not found, channelId: %v", channelId)
+	}
+	if err != nil {
+		return fmt.Errorf("error:%v was seen on retreiving details of channelID:%v",
+			err.Error(), channelId)
+	}
+	if channel.AuthorizedAmount.Cmp(latestAuthorizedAmount) > 0 {
```
[the remark this answers is not in the document]
Comment (vsbogd) [reply escrow/token_service.go]: @anandrgitnirman , sorry, may be it is incorrect comment because I see in code below you allow `AuthorizedAmount` be equal to `latestAuthorizedAmount`. So the question is: what behavior do you expect?

File: escrow/prepaid_service.go [hunk 3]
```diff
@@ -151,13 +151,8 @@ var (
...
 		newState := oldState.Clone()
-		//we dont add the old amount planned amont , we just replace it with the latest planned amount signed
-		newState.UpdateUsageType = PLANNED_AMOUNT
-		newState.PlannedAmount = revisedAmount
-		if newState.PlannedAmount.Cmp(oldState.PlannedAmount) < 0 {
-			return nil, fmt.Errorf("you need to sign for a higher planned amount than %v"+
-				" on the channel Id %v", oldState.PlannedAmount, oldState.ChannelID)
-		}
+		usageKey := &PrePaidDataKey{UsageType: PLANNED_AMOUNT, ChannelID: oldState.ChannelID}
+		updateDetails(newState, usageKey, revisedAmount)
```
[the remark this answers is not in the document]
Comment (anandrgitnirman) [reply escrow/prepaid_service.go:155]: Refund AMount is updated when the service calls fails , we always updte the Used amount assuuming the service calls will succeed.
So lets say planned amount is 3 , used amount = 0 , refund amount = 0  ( you can make 3 calls) 
1 service call is made , but fails, => planned amount = 3, used amount = 1 and refund Amount = 1  
you should be allowed to make 3 more calls 
 new service call 
 2 <= 3 + 1 ( allow) 
 3 <= 3 + 1 ( allow) 
 4 <= 3 + 1  ( allow)  
 5 <= 3 +1 ( Error)

[the remark this answers is not in the document]
Comment (vsbogd) [reply escrow/prepaid_service.go:155]: got it, thanks!

File: escrow/token_service.go (hunk 2 above)
[the remark this answers is not in the document]
Comment (anandrgitnirman) [reply escrow/token_service.go]: hi @vsbogd ,  if you need to renew a token , you dont need to send a Higer signed amount , you can just send the same amount as you last signed

File: snetd/cmd/components.go [hunk 4]
```diff
@@ -545,12 +546,25 @@ func (components *Components) ConfigurationService() *configuration_service.Conf
...
+
+func (components *Components) TokenService() *escrow.TokenService {
 	if components.tokenService != nil {
 		return components.tokenService
 	}
 
-	components.tokenService = token.NewJWTTokenService(*components.OrganizationMetaData())
+	components.tokenService = escrow.NewTokenService(components.PaymentChannelService(),
+		components.PrePaidService(), components.TokenManager(),
+		escrow.NewChannelPaymentValidator(components.Blockchain(), config.Vip(), components.OrganizationMetaData()),
```
[the remark this answers is not in the document]
Comment (anandrgitnirman) [reply snetd/cmd/components.go:566]: #506 , will do this in a seprate PR

File: escrow/memory_storage.go (hunk 1 above)
Comment (anandrgitnirman) [reply escrow/memory_storage.go]: fixed this

File: escrow/memory_storage.go [hunk 5]
```diff
@@ -110,13 +109,104 @@ func (storage *memoryStorage) Clear() (err error) {
...
+			}
+		}
+	}
+	return true, nil
 }
 
 func (client *memoryStorage) ExecuteTransaction(request CASRequest) (ok bool, err error) {
-	return false, fmt.Errorf("Not implemented")
+
+	transaction, err := client.StartTransaction(request.ConditionKeys)
```
Comment (vsbogd) [inline escrow/memory_storage.go:170]: You can do it in separate PR as well, but `ExecuteTransaction` method implementation also could be reused for all kind of storages.

Comment (anandrgitnirman) [reply escrow/memory_storage.go:170]: Sure, let me create a issue on this and will raise a seprate PR , #507
